File: Algorithm/src/main/java/Acwing/Ac896_最长上升子序列2v/Main.java
package Acwing.Ac896_最长上升子序列2v;

import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.io.StreamTokenizer;
import java.util.Stack;

public class Main {
    static BufferedReader in = new BufferedReader(new InputStreamReader(System.in));
    static StreamTokenizer st = new StreamTokenizer(in);
    static int nextInt() throws IOException {
        st.nextToken();
        return (int)st.nval;
    }

    static int getIndex(int[] arr, int target){
        // 找到第一个>=target的数字
        int L = -1,R=arr.length;
        while (L+1!=R){
            int mid = L+R>>1;
            if(arr[mid]>=target){
                R = mid;
            } else{
                L=mid;
            }
        }
        return R;
    }

    static int getIndex2(Stack<Integer> stack, int target){
        int L = -1, R =stack.size();
        while (L+1!=R){
            int mid = L+R>>1;
            if(stack.get(mid)>=target){
                R = mid;
            } else{
                L=mid;
            }
        }
        return R;
    }

    public static void main(String[] args) throws IOException {
        int n = nextInt();
        int[] arr = new int[n];
        for (int i=0;i<n;i++){
            arr[i]=nextInt();
        }
        Stack<Integer> stack = new Stack<>();
        stack.push(arr[0]);
        for(int i=1;i<n;i++){
            if(stack.peek()<arr[i]){
                stack.push(arr[i]);
            } else if (stack.peek() > arr[i]) {
                int index = getIndex2(stack,arr[i]);
                stack.set(index,arr[i]);
            }
        }

        System.out.println(stack.size());
    }
}

File: Algorithm/src/main/java/Acwing/Ac896_最长上升子序列2v/main.go
package main

import (
	"bufio"
	"fmt"
	"os"
	"strconv"
)

type Stack []int

func NewStack() Stack {
	return make(Stack, 0)
}

func (s *Stack) Push(v int) {
	*s = append(*s, v)
}

func (s *Stack) Pop() int {
	ans := (*s)[len(*s)-1]
	*s = (*s)[:len(*s)-1]
	return ans
}

func (s *Stack) Peek() int {
	return (*s)[len(*s)-1]
}

func (s *Stack) Size() int {
	return len(*s)
}

func (s *Stack) Empty() bool {
	return len(*s) == 0
}

var scanner = bufio.NewScanner(os.Stdin)

func nextInt() int {
	scanner.Scan()
	ans, _ := strconv.Atoi(scanner.Text())
	return ans
}
func main() {
	scanner.Split(bufio.ScanWords)
	n := nextInt()
	arr := make([]int, n)
	for i := 0; i < n; i++ {
		arr[i] = nextInt()
	}

	stack := NewStack()
	stack.Push(arr[0])
	for i := 1; i < n; i++ {
		if arr[i] > stack.Peek() {
			stack.Push(arr[i])
		} else if arr[i] < stack.Peek() {
			stack[getIndex(stack, arr[i])] = arr[i]
		}
	}
	fmt.Println(stack.Size())
}
func getIndex(stack Stack, target int) int {
	l, r := -1, stack.Size()
	for l+1 != r {
		mid := (l + r) >> 1
		if stack[mid] >= target {
			r = mid
		} else {
			l = mid
		}
	}
	return r
}
